src/extension/simple_op.rs

```diff
@@ -36,76 +36,99 @@ where
-    fn def_signature(&self) -> SignatureFunc;
+    fn signature(&self) -> SignatureFunc;
 
     /// Description of the operation. By default, the same as `self.name()`.
     fn description(&self) -> String {
         self.name().to_string()
     }
 
-    /// Any type args which define this operation. Default is no type arguments.
-    fn type_args(&self) -> Vec<TypeArg> {
-        vec![]
-    }
-
     /// Edit the opdef before finalising. By default does nothing.
     fn post_opdef(&self, _def: &mut OpDef) {}
 
+    /// Add an operation implemented as an [MakeOpDef], which can provide the data
+    /// required to define an [OpDef], to an extension.
+    fn add_to_extension(&self, extension: &mut Extension) -> Result<(), ExtensionBuildError> {
+        let def = extension.add_op(self.name(), self.description(), self.signature())?;
+
+        self.post_opdef(def);
+
+        Ok(())
+    }
+
+    /// load all variants of a [OpEnum] in to an extension as op defs.
+    fn load_all_ops(extension: &mut Extension) -> Result<(), ExtensionBuildError>
+    where
+        Self: IntoEnumIterator,
+    {
+        for op in Self::iter() {
+            op.add_to_extension(extension)?;
+        }
+        Ok(())
+    }
+}
+
+/// Traits implemented by types which can be loaded from [`ExtensionOp`]s,
+/// i.e. concrete instances of [`OpDef`]s, with defined type arguments.
+pub trait MakeExtensionOp: OpName {
+    /// Try to load one of the operations of this set from an [OpDef].
+    fn from_extension_op(ext_op: &ExtensionOp) -> Result<Self, OpLoadError>
+    where
+        Self: Sized;
     /// Try to instantiate a variant from an [OpType]. Default behaviour assumes
     /// an [ExtensionOp] and loads from the name.
     fn from_optype(op: &OpType) -> Option<Self>
     where
         Self: Sized,
     {
         let ext: &ExtensionOp = op.as_leaf_op()?.as_extension_op()?;
-        Self::from_op_def(ext.def(), ext.args()).ok()
+        Self::from_extension_op(ext).ok()
     }
 
+    /// Any type args which define this operation.
+    fn type_args(&self) -> Vec<TypeArg>;
+
     /// Given the ID of the extension this operation is defined in, and a
     /// registry containing that extension, return a [RegisteredEnum].
     fn to_registered(
         self,
         extension_id: ExtensionId,
         registry: &ExtensionRegistry,
-    ) -> RegisteredEnum<'_, Self>
+    ) -> RegisteredOp<'_, Self>
     where
         Self: Sized,
     {
-        RegisteredEnum {
+        RegisteredOp {
             extension_id,
             registry,
-            op_enum: self,
+            op: self,
         }
     }
+}
 
-    /// Iterator over all operations in the set. Non-trivial variants will have
-    /// default values used for the members.
-    fn all_variants() -> <Self as IntoEnumIterator>::Iterator
+/// Blanket implementation for non-polymorphic operations - no type parameters.
+impl<T: MakeOpDef> MakeExtensionOp for T {
```

Comment:
To be clear, `MakeOpDef` may never have type parameters.
Perhaps mention that in the trait description?

Reply:
they can have _parameters_ defined by their signature (just like OpDefs), but not type arguments